### DISCUSSION TITLE
how to use --symlink commands?

### DISCUSSION THREAD

**Discussion by IssacJun** [oracle/opengrok#4168] [General]
I would like to know how to use --symlink among OpenGrok commands.

I want to use /opengrok/src folder as source root folder and put symlink files inside this folder.
```
/opengrok/src $ ll
total 16
drwxr-xr-x 4 A root         4096  1월 26 15:32 ./
drwxr-xr-x 7 root         root         4096  1월 26 11:14 ../
drwxrwxr-x 9 A A 4096  1월 26 11:58 mustang_1_5_release/
drwxrwxr-x 3 A A 4096  1월 26 14:54 nissan_16.03.31/
lrwxrwxrwx 1 A A   32  1월 26 15:32 symlink -> /home/swut.testbot/workspace/tmp/
```
The symbolic link file points to the /home/A/workspace/tmp folder.

Indexing is done with the command below.
```
java
   -Djava.util.logging.config.file=/opengrok/etc/logging.properties
   -jar /opengrok/dist/lib/opengrok.jar
   -c /usr/local/bin/ctags
   -s /opengrok/src
   -d /opengrok/data
   -H -P -S -G
   -W /opengrok/etc/configuration.xml
   -U http://localhost:8080/source
   --symlink /opengrok/src/symlink
```
But when I connect to OpenGrok, it shows as below.

```
Error: File not found!
The requested resource is not available.

Resource lacks history info. Was remote SCM side up when indexing occurred? Cleanup history cache dir(or just the .gz for the file or db record) and rerun indexer making sure remote side will respond during indexing.
```
What should I do?

**Comment by vladak** [comment]
Symlinks directly underneath source root are automatically allowed, so using the `--symlink` in your case is not needed. What is the request URL that leads to the observed 'File not found' error ?

Also, is there anything relevant/notable in the webapp logs around the time the failed request was made ?

**Reply by nirfuchs73** [reply]
We have the same issue in our OpenGrok server. E.g. we have the following test repository structure, where envsetup.sh file and the target folder are symlink:
```
./envsetup.sh -> /data/opengrok/src/test/build/make/envsetup.sh
./make
./make/envsetup.sh
./make/target
./make/target/Android.mk
./target -> /data/opengrok/src/test/build/make/target/
```

We use the following commands deploy and index:
```
opengrok-deploy -c /opengrok/etc/configuration.xml \
    /opengrok/lib/source.war /usr/local/tomcat/webapps
```
```
opengrok-indexer \
    -J=-Djava.util.logging.config.file=/scripts/logging.properties \
    -a /opengrok/lib/opengrok.jar -- \
    -c /usr/local/bin/ctags \
    -s /opengrok/src -d /opengrok/data -H -P -S -G \
    -W /opengrok/etc/configuration.xml -U http://localhost:8080/source
```
We get the following error when clicking on the envsetup.sh symlink;
```
Error: File not found!
The requested resource is not available.

Resource lacks history info. Was remote SCM side up when indexing occurred? Cleanup history cache dir(or just the .gz for the file or db record) and rerun indexer making sure remote side will respond during indexing.
```
What are we missing?

**Reply by vladak** [reply]
is there anything of interest in the web app logs around the time the error message is displayed in the browser ? what OpenGrok version is this ?

**Reply by nirfuchs73** [reply]
There is no error in the server logs.
I'm running the application in a Docker container. I'm using the following image: opengrok/docker:1.7.32. I tried also the latest version: 1.82 and the behavior is the same.

**Reply by nirfuchs73** [reply]
It seems like symlinks behaves differently inside the Docker container.
I tried it without Docker container. now it is possible to see the file (symlink) content, however when clicking the History we an error:
```
Error: File not found!
The requested resource is not available.
```
The tomcat server show the following warning:
```
WARNING [https-jsse-nio-8443-exec-3] org.opengrok.indexer.history.LatestRevisionUtil.getLatestRevision cannot get latest revision for <path to symlink file>
```

**Comment by nirfuchs73** [comment]
We have the same issue in our OpenGrok server. E.g. we have the following test repository structure, where envsetup.sh file and the target folder are symlink:
./envsetup.sh -> /data/opengrok/src/test/build/make/envsetup.sh
./make
./make/envsetup.sh
./make/target
./make/target/Android.mk
./target -> /data/opengrok/src/test/build/make/target/

We use the following commands deploy and index:
opengrok-deploy -c /opengrok/etc/configuration.xml /opengrok/lib/source.war /usr/local/tomcat/webapps

opengrok-indexer \
    -J=-Djava.util.logging.config.file=/scripts/logging.properties \
    -a /opengrok/lib/opengrok.jar -- \
    -c /usr/local/bin/ctags \
    -s /opengrok/src -d /opengrok/data -H -P -S -G \
    -W /opengrok/etc/configuration.xml -U http://localhost:8080/source
 
We get the following error when clicking on the envsetup.sh symlink;
Error: File not found!
The requested resource is not available.

Resource lacks history info. Was remote SCM side up when indexing occurred? Cleanup history cache dir(or just the .gz for the file or db record) and rerun indexer making sure remote side will respond during indexing.

What are we missing?
